### DISCUSSION TITLE
"document is not defined" - client-side javascript with `useEffect`

### DISCUSSION THREAD

**Discussion by kr8n3r** [vercel/next.js#14071] [Help] [answered]
Trying to utilise a client-side script from a UI framework results in "document is not defined"
It also seems like a basic console.log doesn't show up either

**Note: simple CRA app with the same component running the client script just fine: https://github.com/kr8n3r/simple-cra**
<details>
<summary>Toggle in action</summary>

![app](https://user-images.githubusercontent.com/3758555/84398376-c5ade080-abf7-11ea-8eb5-0e9531eee095.gif)

</details>

Current setup below (or [PR in progress](https://github.com/alphagov/paas-product-pages/pull/25))
```
import React, {useEffect, useRef} from "react";
import HeaderJS from 'govuk-frontend/govuk/components/header/header';

const Header = () => {
  const headerRef = useRef();

  useEffect(() => {
    new HeaderJS(headerRef.current).init();
  }, [headerRef]);

  return (
    <header className="header" 
            role="banner" 
            data-module="header" 
            ref={headerRef}>
          <button type="button" className="js-header-toggle">Menu</button>
         <nav>
            <ul id="navigation">.....</ul>
         </nav>
    </header>
  )
}
export default Header
```
Package script in question: https://github.com/alphagov/govuk-frontend/blob/master/package/govuk/components/header/header.js
Full error
```
ReferenceError: document is not defined
    at /App/node_modules/govuk-frontend/govuk/components/header/header.js:291:39
    at /App/node_modules/govuk-frontend/govuk/components/header/header.js:298:2
    at detect.defineProperty.a (/App/node_modules/govuk-frontend/govuk/components/header/header.js:2:82)
    at Object.<anonymous> (/App/node_modules/govuk-frontend/govuk/components/header/header.js:5:2)
    at Module._compile (internal/modules/cjs/loader.js:701:30)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:712:10)
    at Module.load (internal/modules/cjs/loader.js:600:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:539:12)
    at Function.Module._load (internal/modules/cjs/loader.js:531:3)
    at Module.require (internal/modules/cjs/loader.js:637:17)
    at require (internal/modules/cjs/helpers.js:22:18)
    at eval (webpack-internal:///govuk-frontend/govuk/components/header/header:1:18)
    at Object.govuk-frontend/govuk/components/header/header (/App/.next/server/static/development/pages/_document.js:252:1)
    at __webpack_require__ (/App/.next/server/static/development/pages/_document.js:23:31)
    at eval (webpack-internal:///./components/Header/index.tsx:4:103)
    at Module../components/Header/index.tsx (/App/.next/server/static/development/pages/_document.js:182:1)
```

**Comment by jamesmosier** [comment, marked as answer]
You'll have to wait for the client to render this, as `document` is not available on the server or when this page is built.

You could conditionally load this component by checking if `typeof window !== 'undefined'` then render the component.

As a last resort you could import this component using a dynamic import with `srr: false` [as shown in the docs](https://nextjs.org/docs/advanced-features/dynamic-import#with-no-ssr)

---

You could do something like this to conditionally render a component:

```jsx
// _app.js
function App({Component}) {
  return (
    <>
      <HeaderImportedFromApp />
      <Component/>
    </>
  )
}

// HeaderImportedFromApp.js
function HeaderImportedFromApp() {
  if (typeof window === 'undefined') {
    return null
  }

  return <HeaderWithDocumentIssue />
}

// HeaderWithDocumentIssue.js
import HeaderJS from 'govuk-frontend/govuk/components/header/header';
function HeaderWithDocumentIssue() {
  // use HeaderJS since window is defined
}
```

**Reply by todortotev** [reply]
Just out of curiosity, why would a "last resort" be the dynamic import?

**Reply by jamesmosier** [reply]
Good question! I guess "last resort" is a bit strong. I just usually view these kinds of problems as easier to just do a conditional, checking for `typeof window`, so then you don't have to worry about doing a dynamic import anytime you encounter a component that uses `window`/`document`.

**Reply by severin2** [reply]
> Just out of curiosity, why would a "last resort" be the dynamic import?

other tooling that we'd want to just work alongside the app, EX jest, won't know what is going on with the dynamic import, for one.

**Reply by brogr** [reply]
This doesn't work in cases where the error is already caused by importing HeaderWithDocumentIssue, which in turn imports 'govuk-frontend/govuk/components/header/header':

`// HeaderImportedFromApp.js
import HeaderWithDocumentIssue from "./HeaderWithDocumentIssue";
function HeaderImportedFromApp() {
  if (typeof window === 'undefined') {
    return null
  }

  return <HeaderWithDocumentIssue />
}`

**Comment by kr8n3r** [comment]
Thanks @jamesmosier. Not sure this is going to work in _document.js

**Reply by kr8n3r** [reply]
a reduced example of 
```
useEffect(() => {
    const isBrowser = typeof window !== 'undefined'
    if (isBrowser) {
      console.log('red')
    }
  }, [headerRef]);
```
doesn't produce anything

**Reply by jamesmosier** [reply]
FWIW I wouldn't personally put my `<header>` component inside `_document.js`, I would put it within `_app.js`. 

`_document` is different than all other pages so your `useEffect` won't run as you want in there. [`Document` is only rendered in the server](https://nextjs.org/docs/advanced-features/custom-document#caveats). You'll have to move it to `_app`.

**Reply by kr8n3r** [reply]
ok, but document is needed to have a custom HTML markup (lang param, html and body classes, etc) - i guess use react-helmet?

Unfortunately, even moving the header to app.js still results in document not defined error (with or without type window check)

**Reply by jamesmosier** [reply]
You can keep `_document`, you just can't have `useEffect` run in there.

This is what you would have to do:

```jsx
// _app.js
function App({Component}) {
  return (
    <>
      <HeaderImportedFromApp />
      <Component/>
    </>
  )
}

// HeaderImportedFromApp.js
function HeaderImportedFromApp() {
  if (typeof window === 'undefined') {
    return null
  }

  return <HeaderWithDocumentIssue />
}

// HeaderWithDocumentIssue.js
import HeaderJS from 'govuk-frontend/govuk/components/header/header';
function HeaderWithDocumentIssue() {
  // use HeaderJS since window is defined
}
```

The import by itself is the issue because that component does not protect against the possibility that document/window may not exist. So you have to avoid importing that component altogether within Next while document/window are undefined.

**Reply by kr8n3r** [reply]
ok, I don't have `useEffect` in _document, it's in `Header`

If i moved the Header import from _document to _app, I still get "document not defined"

**Reply by jamesmosier** [reply]
I understand 😄  you will **always get that error** whether it is in `_document` or `_app` or even a `pages/*` file. Next compiles pages on the server in a Node environment.

So you must work around it via something like [my answer above](https://github.com/vercel/next.js/discussions/14071#discussioncomment-25099) or by not rendering that component until `window` is defined, as I showed in my previous comment in this thread.

**Reply by kr8n3r** [reply]
cheers James.

can we bypass that by moving initialisation to an external script file in a separate bundle (configured in next.config)?

**Reply by jamesmosier** [reply]
That I'm not sure of.

**Reply by kr8n3r** [reply]
i can try if we have an example of nextjs.config that creates a new bundle a new entry and if `_app.js` can accept as script being placed anywhere

**Reply by jamesmosier** [reply]
I don't think there is any straightforward way to do that. You'd be on your own as far as writing custom webpack config logic and even then I'm not sure what is possible.

**Reply by kr8n3r** [reply]
I mean, I have no option if I need to load external javascript that handles component interactivity

What is confuses me is that the following in `_app.js` runs fine serverside, but when `document` is used by the imported script, `document` is not defined
```
  useEffect(() => {
    document.documentElement.classList.add('govuk-template');
    document.body.classList.add('js-enabled', 'govuk-template__body');
  }, []);
```

Ideally I'd want to SSR the Header as interactivity can kick in with a delay

thanks for the insight so far

**Comment by oyilmaztekin** [comment]
I think, dynamic import can be used here. 

```js
useEffect(async () => {
  if (typeof window !== "undefined") {
    const { aLibraryUseDocument } = await import("plugin");
    aLibraryUseDocument.method();
  }
});
```

**Reply by Galionix** [reply]
how can you use await here? useEffect cannot be given async function

**Reply by oyilmaztekin** [reply]
it's my bad. I think I had missed to add `async` expression just before the callback. I have fixed it now.
